pkg/endpointmanager/manager.go

```diff
@@ -301,16 +305,28 @@ func (mgr *EndpointManager) ReleaseID(ep *endpoint.Endpoint) error {
 // Unexpose removes the endpoint from the EndpointManager so that other
 // subsystems will no longer find it during future lookups. The endpoint may
 // still be in use from other subsystems if it was found prior to this call.
-//
-// Must be called while holding the Endpoint lock for reading.
 func (mgr *EndpointManager) Unexpose(ep *endpoint.Endpoint) chan struct{} {
 	epRemoved := make(chan struct{})
 
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	// Fetch the identifiers; this will only fail if the endpoint is
+	// already disconnected, in which case we don't need to proceed with
+	// the rest of cleaning up the endpoint.
+	identifiers, err := ep.Identifiers()
+	if err != nil {
+		// Already disconnecting
+		close(epRemoved)
+		return epRemoved
+	}
+	previousState := ep.Disconnect()
```

Comment:
The goal at that point of the PR was to prevent other calls coming into `endpointManager.RemoveEndpoint()` and concurrently attempting to fetch the identifiers / Disconnect the endpoint.

If two goroutines were to reach this point at the same time, then both could do `ep.Identifiers()` and succeed, then both attempt to Disconnect / clean up the endpoint. If instead the mutex covers this section, then one will grab the identifiers, disconnect the endpoint & clean up; then the second call would come in, attempt to get the identifiers and fail (because the endpoint was already disconnected), then exit out.

Though that said, I think that the rest of this function is idempotent so it's probably not such a big deal. Maybe we could reduce the size of the critical section here and reduce interaction between the `EndpointManager` mutex and the `Endpoint` mutex. Though that said, I would expect if both mutexes are held at any point in the code, then they should be held in that order (first grab `EndpointManager` mutex and then hold `Endpoint` mutex), to prevent deadlocks.

Reply:
Thank you for taking the time to explain! The piece I was missing was that `ep.Identifiers` will fail in case the endpoint was already disconnected. Makes perfect sense this way.